File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveUploadSession.java
package ru.kmorozov.onedrive.client;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Arrays;

public class OneDriveUploadSession {

    // Upload in chunks of 5MB as per MS recommendation
    private static final int CHUNK_SIZE = 5 * 1024 * 1024;
    private final File file;
    private final String uploadUrl;
    private final RandomAccessFile raf;
    private final OneDriveItem parent;
    private OneDriveUploadSession.Range[] ranges;
    private long totalUploaded;
    private long lastUploaded;
    private OneDriveItem item;

    public OneDriveUploadSession(OneDriveItem parent, File file, String uploadUrl, String[] ranges) throws IOException {
        this.parent = parent;
        this.file = file;
        this.uploadUrl = uploadUrl;
        raf = new RandomAccessFile(file, "r");
        this.setRanges(ranges);
    }

    public void setRanges(String[] stringRanges) {

        ranges = new OneDriveUploadSession.Range[stringRanges.length];
        for (int i = 0; i < stringRanges.length; i++) {
            long start = Long.parseLong(stringRanges[i].substring(0, stringRanges[i].indexOf('-')));

            String s = stringRanges[i].substring(stringRanges[i].indexOf('-') + 1);

            long end = 0L;
            if (!s.isEmpty()) {
                end = Long.parseLong(s);
            }

            this.ranges[i] = new OneDriveUploadSession.Range(start, end);
        }

        if (0 < this.ranges.length) {
            this.lastUploaded = this.ranges[0].start - this.totalUploaded;
            this.totalUploaded = this.ranges[0].start;
        }
    }

    public byte[] getChunk() throws IOException {

        byte[] bytes = new byte[OneDriveUploadSession.CHUNK_SIZE];

        this.raf.seek(this.totalUploaded);
        int read = this.raf.read(bytes);

        if (OneDriveUploadSession.CHUNK_SIZE > read) {
            bytes = Arrays.copyOf(bytes, read);
        }

        return bytes;
    }

    public long getTotalUploaded() {
        return this.totalUploaded;
    }

    public long getLastUploaded() {
        return this.lastUploaded;
    }

    public OneDriveItem getParent() {
        return this.parent;
    }

    public String getUploadUrl() {
        return this.uploadUrl;
    }

    public File getFile() {
        return this.file;
    }

    public boolean isComplete() {
        return null != this.item;
    }

    public void setComplete(OneDriveItem item) {
        this.item = item;
        this.lastUploaded = this.file.length() - this.totalUploaded;
        this.totalUploaded = this.file.length();
    }

    public OneDriveItem getItem() {
        return this.item;
    }

    private static final class Range {
        public long start;
        public long end;

        private Range(long start, long end) {
            this.start = start;
            this.end = end;
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveUploadSession.kt
package ru.kmorozov.onedrive.client

import java.io.File
import java.io.IOException
import java.io.RandomAccessFile
import java.util.*

class OneDriveUploadSession @Throws(IOException::class)
constructor(val parent: OneDriveItem, val file: File, val uploadUrl: String, ranges: Array<String>) {
    private val raf: RandomAccessFile
    private var ranges: Array<Range?>? = null
    var totalUploaded: Long = 0
        private set
    var lastUploaded: Long = 0
        private set
    var item: OneDriveItem? = null
        private set

    val chunk: ByteArray
        @Throws(IOException::class)
        get() {

            var bytes = ByteArray(CHUNK_SIZE)

            raf.seek(totalUploaded)
            val read = raf.read(bytes)

            if (CHUNK_SIZE > read) {
                bytes = Arrays.copyOf(bytes, read)
            }

            return bytes
        }

    val isComplete: Boolean
        get() = null != item

    init {
        this.raf = RandomAccessFile(file, "r")
        setRanges(ranges)
    }

    fun setRanges(stringRanges: Array<String>) {

        this.ranges = arrayOfNulls(stringRanges.size)
        for (i in stringRanges.indices) {
            val start = java.lang.Long.parseLong(stringRanges[i].substring(0, stringRanges[i].indexOf('-')))

            val s = stringRanges[i].substring(stringRanges[i].indexOf('-') + 1)

            var end = 0L
            if (!s.isEmpty()) {
                end = java.lang.Long.parseLong(s)
            }

            ranges!![i] = Range(start, end)
        }

        if (ranges!!.isNotEmpty()) {
            lastUploaded = ranges!![0]!!.start - totalUploaded
            totalUploaded = ranges!![0]!!.start
        }
    }

    fun setComplete(item: OneDriveItem) {
        this.item = item
        lastUploaded = file.length() - totalUploaded
        totalUploaded = file.length()
    }

    private class Range(var start: Long, var end: Long)

    companion object {

        // Upload in chunks of 5MB as per MS recommendation
        private const val CHUNK_SIZE = 5 * 1024 * 1024
    }
}
